clickhouse 13.3-RELEASE FreeBSD 'c++: error: no such file or directory: 'libclang_rt.builtins-x86_64.a'

Hello,
 I try install clickhouse in Freebsd Jail, but couldn't compile it from source

Cmake options (-DCMAKE_CXX_STANDARD=20 -DENABLE_EMBEDDED_COMPILER=OFF -DENABLE_THINLTO=OFF -DENABLE_PARQUET=OFF -DENABLE_UTILS=ON -DNO_WERROR=1 -DCOMPILER_CACHE=disabled; ) I take from  /usr/ports/databases/clickhouse/Makefile

**maybe someone have correct instruction how compile clickhouse on freebsd ?**


-- Git HEAD commit hash: e3c09e97034b7d26b2280487fa8422a8037caa51


> Make sure that `git diff` result is empty and you've just pulled fresh master. Try cleaning up cmake cache. Just in case, official build instructions are published here: https://clickhouse.com/docs/en/development/build/

Operating system: FreeBSD 13.3-RELEASE 

> OS kind or distribution, specific version/release, non-standard kernel if any. If you are trying to build inside virtual machine, please mention it too.

make version: cmake-3.28.1 

Ninja version: ninja-1.11.1,2  

Compiler name and version: clang version 17.0.6

**Full cmake and/or ninja output**




git clone --recursive --shallow-submodules https://github.com/ClickHouse/ClickHouse.git
cd ClickHouse; \
mkdir build; \
```
cmake -S . -B build \
-DCMAKE_CXX_STANDARD=20 \
-DENABLE_EMBEDDED_COMPILER=OFF \
-DENABLE_THINLTO=OFF \
-DENABLE_PARQUET=OFF \
-DENABLE_UTILS=ON \
-DNO_WERROR=1 \
-DCOMPILER_CACHE=disabled; \
-- The C compiler identification is Clang 17.0.6
-- The CXX compiler identification is Clang 17.0.6
-- The ASM compiler identification is Clang with GNU-like command-line
-- Found assembler: /usr/bin/cc
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using compiler:
FreeBSD clang version 17.0.6 (https://github.com/llvm/llvm-project.git llvmorg-17.0.6-0-g6009708b4367)
Target: x86_64-unknown-freebsd13.3
Thread model: posix
InstalledDir: /usr/bin
-- Using linker: <default>
-- Using archiver: /usr/bin/llvm-ar
-- Using ranlib: /usr/bin/llvm-ranlib
-- Using install-name-tool: /usr/local/bin/llvm-install-name-tool
-- Using objcopy: /usr/bin/llvm-objcopy
-- Using strip: /usr/bin/llvm-strip
-- Using *ccache: no (disabled via configuration)
-- Found Git: /usr/local/bin/git (found version "2.43.0") 
-- Git HEAD commit hash: e3c09e97034b7d26b2280487fa8422a8037caa51
-- CMAKE_BUILD_TYPE is not set, set to default = RelWithDebInfo                                                                                 
-- CMAKE_BUILD_TYPE: RelWithDebInfo                                                                                                             
-- Performing Test SUPPORTS_CXXFLAG_no_enum_constexpr_conversion        
-- Performing Test SUPPORTS_CXXFLAG_no_enum_constexpr_conversion - Success                                                                      
-- Performing Test SUPPORTS_CFLAG_no_enum_constexpr_conversion                                                                                                                                                                                                                                  -- Performing Test SUPPORTS_CFLAG_no_enum_constexpr_conversion - Success                                                                                                                                                                                                                        -- No official build: A checksum hash will not be added to the clickhouse executable                                                                                                                                                                                                            -- Default libraries: -nodefaultlibs libclang_rt.builtins-x86_64.a  -lc -lm -lrt -lpthread                                                                                                                                                                                                      
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD                                                                                                                                                                                                                                                      -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed                                                                                                                                                                                                                                             -- Check if compiler accepts -pthread                                                                                                                                                                                                                                                           -- Check if compiler accepts -pthread - yes                                                                                                                                                                                                                                                     
-- Found Threads: TRUE                                               
-- Unit tests are enabled                                                                                                                       
-- Building for: FreeBSD-13.3-RELEASE amd64                                                                                                     
-- Adding contrib module boringssl (configuring with boringssl-cmake)                                                                           
-- Adding contrib module miniselect (configuring with miniselect-cmake)                                                                         
-- Adding contrib module pdqsort (configuring with pdqsort-cmake)                                                                               
-- Adding contrib module pocketfft (configuring with pocketfft-cmake)                                                                           
-- Adding contrib module crc32-vpmsum (configuring with crc32-vpmsum-cmake)                                                                     
-- crc32-vpmsum library is only supported on ppc64le                                                                                            
-- Adding contrib module sparsehash-c11 (configuring with sparsehash-c11-cmake)                                                                 
-- Adding contrib module abseil-cpp (configuring with abseil-cpp-cmake)                                                                         
-- Adding contrib module magic_enum (configuring with magic-enum-cmake)                                                                         
-- Adding contrib module boost (configuring with boost-cmake)    
-- Adding contrib module cctz (configuring with cctz-cmake)             
-- Packaging with tzdata version: 2024a                                 
-- Adding contrib module consistent-hashing (configuring with consistent-hashing)                                                               
-- Adding contrib module dragonbox (configuring with dragonbox-cmake)                                                                           
-- Adding contrib module vectorscan (configuring with vectorscan-cmake)
-- Adding contrib module jemalloc (configuring with jemalloc-cmake)     
CMake Warning at contrib/jemalloc-cmake/CMakeLists.txt:20 (message):                                                                            
  jemalloc support on non-Linux is EXPERIMENTAL   
-- jemalloc malloc_conf: oversize_threshold:0,muzzy_decay_ms:0,dirty_decay_ms:5000                                                              
-- Adding contrib module libcpuid (configuring with libcpuid-cmake)     
-- Adding contrib module libdivide-cmake (configuring with libdivide-cmake)                                                                                                                                                                                                                     
-- Adding contrib module libmetrohash (configuring with libmetrohash)
-- Adding contrib module lz4 (configuring with lz4-cmake)                                                                                       
-- Adding contrib module murmurhash (configuring with murmurhash)       
-- Adding contrib module replxx (configuring with replxx-cmake)      
-- Adding contrib module unixodbc (configuring with unixodbc-cmake)  
-- ODBC is only supported on Linux with dynamic linking               
-- Not using ODBC                                                       
-- Adding contrib module nanodbc (configuring with nanodbc-cmake)  
-- Adding contrib module capnproto (configuring with capnproto-cmake)   
-- Adding contrib module yaml-cpp (configuring with yaml-cpp-cmake)                                                                             
-- Adding contrib module re2 (configuring with re2-cmake)                                                                                       
-- Adding contrib module xz (configuring with xz-cmake)               
-- Adding contrib module brotli (configuring with brotli-cmake)                                                                                 
-- Adding contrib module double-conversion (configuring with double-conversion-cmake)                                                           
-- Adding contrib module croaring (configuring with croaring-cmake)
-- Adding contrib module zstd (configuring with zstd-cmake)
-- Adding contrib module zlib-ng (configuring with zlib-ng-cmake)
-- Adding contrib module bzip2 (configuring with bzip2-cmake)
-- Adding contrib module minizip-ng (configuring with minizip-ng-cmake)
-- Adding contrib module snappy (configuring with snappy-cmake)
-- Adding contrib module rocksdb (configuring with rocksdb-cmake)
-- Adding contrib module thrift (configuring with thrift-cmake)
-- Adding contrib module arrow (configuring with arrow-cmake)
-- Not using parquet
-- Adding contrib module avro (configuring with avro-cmake)
-- Adding contrib module google-protobuf (configuring with google-protobuf-cmake)
-- Adding contrib module openldap (configuring with openldap-cmake)
-- Adding contrib module grpc (configuring with grpc-cmake)
-- Adding contrib module msgpack-c (configuring with msgpack-c-cmake)
-- Adding contrib module libarchive (configuring with libarchive-cmake)
-- Adding contrib module corrosion (configuring with corrosion-cmake)
no such file or directory
CMake Warning at contrib/corrosion/cmake/FindRust.cmake:176 (message):
  The rustc executable was not found.  Rust not installed or ~/.cargo/bin not
  added to path? Aborting further actions of find_package(Rust).
Call Stack (most recent call first):
  contrib/corrosion-cmake/CMakeLists.txt:5 (find_package)

-- Not using rust
-- Adding contrib module wyhash (configuring with wyhash-cmake)
-- Adding contrib module cityhash102 (configuring with cityhash102)
-- Adding contrib module libfarmhash (configuring with libfarmhash)
-- Adding contrib module icu (configuring with icu-cmake)
-- Not using icu
-- Adding contrib module h3 (configuring with h3-cmake)
-- Adding contrib module mariadb-connector-c (configuring with mariadb-connector-c-cmake)
-- Build without mysqlclient (support for MYSQL dictionary source will be disabled)
-- Adding contrib module libfiu (configuring with libfiu-cmake)
-- Adding contrib module googletest (configuring with googletest-cmake)
-- Adding contrib module llvm-project (configuring with llvm-project-cmake)
-- LLVM TARGETS TO BUILD X86
-- LLVM CMAKE CROSS COMPILING 0
CMake Deprecation Warning at contrib/llvm-project/llvm/CMakeLists.txt:8 (cmake_policy):
  The OLD behavior for policy CMP0114 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


CMake Deprecation Warning at contrib/llvm-project/llvm/CMakeLists.txt:13 (cmake_policy):
  The OLD behavior for policy CMP0116 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- Looking for dlfcn.h
-- Looking for dlfcn.h - found
-- Looking for errno.h
-- Looking for errno.h - found
-- Looking for fcntl.h
-- Looking for fcntl.h - found
-- Looking for link.h
-- Looking for link.h - found
-- Looking for malloc/malloc.h
-- Looking for malloc/malloc.h - not found
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for signal.h
-- Looking for signal.h - found
-- Looking for sys/ioctl.h
-- Looking for sys/ioctl.h - found
-- Looking for sys/mman.h
-- Looking for sys/mman.h - found
-- Looking for sys/param.h
-- Looking for sys/param.h - found
-- Looking for sys/resource.h
-- Looking for sys/resource.h - found
-- Looking for sys/stat.h
-- Looking for sys/stat.h - found
-- Looking for sys/time.h
-- Looking for sys/time.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for sysexits.h
-- Looking for sysexits.h - found
-- Looking for termios.h
-- Looking for termios.h - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for valgrind/valgrind.h
-- Looking for valgrind/valgrind.h - not found
-- Looking for fenv.h
-- Looking for fenv.h - found
-- Looking for FE_ALL_EXCEPT
-- Looking for FE_ALL_EXCEPT - found
-- Looking for FE_INEXACT
-- Looking for FE_INEXACT - found
-- Looking for mach/mach.h
-- Looking for mach/mach.h - not found
-- Looking for CrashReporterClient.h
-- Looking for CrashReporterClient.h - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Looking for pthread_rwlock_init in pthread
-- Looking for pthread_rwlock_init in pthread - found
-- Looking for pthread_mutex_lock in pthread
-- Looking for pthread_mutex_lock in pthread - found
-- Looking for dlopen in dl
-- Looking for dlopen in dl - found
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Looking for xar_open in xar
-- Looking for xar_open in xar - not found
-- Looking for arc4random
-- Looking for arc4random - found
-- Performing Test C_SUPPORTS_WERROR_UNGUARDED_AVAILABILITY_NEW
-- Performing Test C_SUPPORTS_WERROR_UNGUARDED_AVAILABILITY_NEW - Success
-- Looking for __register_frame
-- Looking for __register_frame - found
-- Looking for __deregister_frame
-- Looking for __deregister_frame - found
-- Looking for __unw_add_dynamic_fde
-- Looking for __unw_add_dynamic_fde - not found
-- Looking for _Unwind_Backtrace
-- Looking for _Unwind_Backtrace - found
-- Looking for getpagesize
-- Looking for getpagesize - found
-- Looking for sysconf
-- Looking for sysconf - found
-- Looking for getrusage
-- Looking for getrusage - found
-- Looking for setrlimit
-- Looking for setrlimit - found
-- Looking for isatty
-- Looking for isatty - found
-- Looking for futimens
-- Looking for futimens - found
-- Looking for futimes
-- Looking for futimes - found
-- Looking for sigaltstack
-- Looking for sigaltstack - found
-- Looking for lseek64
-- Looking for lseek64 - not found
-- Looking for mallctl
-- Looking for mallctl - found
-- Looking for mallinfo
-- Looking for mallinfo - not found
-- Looking for mallinfo2
-- Looking for mallinfo2 - not found
-- Looking for malloc_zone_statistics
-- Looking for malloc_zone_statistics - not found
-- Looking for getrlimit
-- Looking for getrlimit - found
-- Looking for posix_spawn
-- Looking for posix_spawn - found
-- Looking for pread
-- Looking for pread - found
-- Looking for sbrk
-- Looking for sbrk - found
-- Looking for strerror
-- Looking for strerror - found
-- Looking for strerror_r
-- Looking for strerror_r - found
-- Looking for strerror_s
-- Looking for strerror_s - not found
-- Looking for setenv
-- Looking for setenv - found
-- Looking for dlopen
-- Looking for dlopen - found
-- Looking for dladdr
-- Looking for dladdr - found
-- Performing Test HAVE_STRUCT_STAT_ST_MTIMESPEC_TV_NSEC
-- Performing Test HAVE_STRUCT_STAT_ST_MTIMESPEC_TV_NSEC - Success
-- Performing Test HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC
-- Performing Test HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC - Success
-- Performing Test HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC
-- Performing Test HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC - Success
-- Looking for __GLIBC__
-- Looking for __GLIBC__ - not found
-- Looking for pthread_getname_np
-- Looking for pthread_getname_np - found
-- Looking for pthread_setname_np
-- Looking for pthread_setname_np - found
-- Looking for proc_pid_rusage
-- Looking for proc_pid_rusage - not found
-- Performing Test HAVE_STD_IS_TRIVIALLY_COPYABLE
-- Performing Test HAVE_STD_IS_TRIVIALLY_COPYABLE - Success
-- Performing Test SUPPORTS_NO_PIE_FLAG
-- Performing Test SUPPORTS_NO_PIE_FLAG - Success
-- Performing Test SUPPORTS_VARIADIC_MACROS_FLAG
-- Performing Test SUPPORTS_VARIADIC_MACROS_FLAG - Success
-- Performing Test SUPPORTS_GNU_ZERO_VARIADIC_MACRO_ARGUMENTS_FLAG
-- Performing Test SUPPORTS_GNU_ZERO_VARIADIC_MACRO_ARGUMENTS_FLAG - Success
-- Native target architecture is X86
-- Threads enabled.
-- Doxygen disabled.
-- Go bindings disabled.
-- Ninja version: 1.11.1
-- Could NOT find OCaml (missing: OCAMLFIND OCAML_VERSION OCAML_STDLIB_PATH) 
-- OCaml bindings disabled.
-- Could NOT find Python module pygments
-- Could NOT find Python module pygments.lexers.c_cpp
-- Could NOT find Python module yaml
-- LLVM host triple: x86_64-unknown-freebsd13.3
-- LLVM default target triple: x86_64-unknown-freebsd13.3
-- Performing Test SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG - Success                                                                                                                                                                                                                 [230/9774]
-- Performing Test C_SUPPORTS_WERROR_DATE_TIME
-- Performing Test C_SUPPORTS_WERROR_DATE_TIME - Success
-- Performing Test CXX_SUPPORTS_WERROR_DATE_TIME
-- Performing Test CXX_SUPPORTS_WERROR_DATE_TIME - Success
-- Performing Test CXX_SUPPORTS_WERROR_UNGUARDED_AVAILABILITY_NEW
-- Performing Test CXX_SUPPORTS_WERROR_UNGUARDED_AVAILABILITY_NEW - Success
-- Performing Test CXX_SUPPORTS_MISSING_FIELD_INITIALIZERS_FLAG
-- Performing Test CXX_SUPPORTS_MISSING_FIELD_INITIALIZERS_FLAG - Success
-- Performing Test C_SUPPORTS_CXX98_COMPAT_EXTRA_SEMI_FLAG
-- Performing Test C_SUPPORTS_CXX98_COMPAT_EXTRA_SEMI_FLAG - Success
-- Performing Test CXX_SUPPORTS_CXX98_COMPAT_EXTRA_SEMI_FLAG
-- Performing Test CXX_SUPPORTS_CXX98_COMPAT_EXTRA_SEMI_FLAG - Success
-- Performing Test C_SUPPORTS_IMPLICIT_FALLTHROUGH_FLAG
-- Performing Test C_SUPPORTS_IMPLICIT_FALLTHROUGH_FLAG - Success
-- Performing Test CXX_SUPPORTS_IMPLICIT_FALLTHROUGH_FLAG
-- Performing Test CXX_SUPPORTS_IMPLICIT_FALLTHROUGH_FLAG - Success
-- Performing Test C_SUPPORTS_COVERED_SWITCH_DEFAULT_FLAG
-- Performing Test C_SUPPORTS_COVERED_SWITCH_DEFAULT_FLAG - Success
-- Performing Test CXX_SUPPORTS_COVERED_SWITCH_DEFAULT_FLAG
-- Performing Test CXX_SUPPORTS_COVERED_SWITCH_DEFAULT_FLAG - Success
-- Performing Test CXX_SUPPORTS_CLASS_MEMACCESS_FLAG
-- Performing Test CXX_SUPPORTS_CLASS_MEMACCESS_FLAG - Success
-- Performing Test CXX_SUPPORTS_NOEXCEPT_TYPE_FLAG
-- Performing Test CXX_SUPPORTS_NOEXCEPT_TYPE_FLAG - Success
-- Performing Test CXX_WONT_WARN_ON_FINAL_NONVIRTUALDTOR
....
-- Performing Test SUPPORTS_CFLAG_no_switch_default                                                                                                                                                                                                                                             
-- Performing Test SUPPORTS_CFLAG_no_switch_default - Success                                                                                                                                                                                                                                   
-- compiler C   = /usr/bin/cc  -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -ffile-prefix-map=/usr/work/ClickHouse=. -falign-functions=32 -mbranches-with
in-32B-boundaries -ffp-contract=off  -fdiagnostics-absolute-paths -Wframe-larger-than=65536 -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-sign-conversion -Wno-implicit-int-conversion -Wno-implicit-int-float-conversion -
Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-fu
nc-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -Wno-enum-constexpr-conversion -Wno-unsafe-buffer-usage -Wno-switch-default -O2 -g -DNDEBUG -O3 -g                                                                           
-- compiler CXX = /usr/bin/c++  -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion -fsized-deallocation  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -ffile-prefix-map=/usr/work/ClickHouse=. -falign-functi
ons=32 -mbranches-within-32B-boundaries -ffp-contract=off  -fdiagnostics-absolute-paths -fstrict-vtable-pointers -Wall -Wextra -Wframe-larger-than=65536 -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-sign-conversion -Wno
-implicit-int-conversion -Wno-implicit-int-float-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variabl
e-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -Wno-enum-constexpr-conversion -Wno-unsafe-buffer-usage -Wno-switch-default -O2 -g -DNDEBUG -O3 -g              
-- LINKER_FLAGS =                                                                                                                                                                                                                                                                               
-- Using Poco::Crypto                                                                                                                                                                                                                                                                           
-- Not using Poco::Data::ODBC                                                                                                                                                                                                                                                                   
The auto-calculated compile jobs limit (1) underutilizes CPU cores (56). Set PARALLEL_COMPILE_JOBS to override.                                                                                                                                                                                 
The auto-calculated link jobs limit (1) underutilizes CPU cores (56). Set PARALLEL_LINK_JOBS to override.                                                                                                                                                                                       
-- Building sub-tree with 1 compile jobs and 1 linker jobs (system: 56 cores, 0 MB RAM, 'OFF' means the native core count).                                                                                                                                                                     
-- Will build ClickHouse 24.4.1.1 revision 54485                                                                                                                                                                                                                                                
-- StorageFileLog is only supported on Linux                                                                                                                                                                                                                                                    
-- clickhouse-keeper and clickhouse-keeper-converter will not be built (lack of NuRaft)                                                                                                                                                                                                         
-- ClickHouse extra components:                                                                                                                                                                                                                                                                 
-- Self-extracting executable: OFF                                                                                                                                                                                                                                                              
-- ODBC bridge mode: OFF                                                                                                                                                                                                                                                                        
-- Library bridge mode: ON                                                                                                                                                                                                                                                                      
-- ClickHouse keeper mode: OFF                                                                                                                                                                                                                                                                  
-- ClickHouse keeper-converter mode: OFF                                                                                                                                                                                                                                                        
-- ClickHouse keeper-client mode: ON                                                                                                                                                                                                                                                            
-- bash_completion will be written to /usr/local/share/bash-completion/completions                                                                                                                                                                                                              
-- Adding alias chc for clickhouse-client                                                                                                                                                                                                                                                       
-- Adding alias chl for clickhouse-local                                                                                                                                                                                                                                                        
-- Adding alias ch for clickhouse-local                                                                                                                                                                                                                                                         
The auto-calculated link jobs limit (1) underutilizes CPU cores (56). Set PARALLEL_LINK_JOBS to override.                                                                                                                                                                                       
-- Building sub-tree with OFF compile jobs and 1 linker jobs (system: 56 cores, 0 MB RAM, 'OFF' means the native core count).                                                                                                                                                                   
CMake Warning at utils/keeper-data-dumper/CMakeLists.txt:2 (message):                                                                                                                                                                                                                           
  Not building keeper-data-dumper due to nuraft is disabled                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                
CMake Warning at utils/memcpy-bench/CMakeLists.txt:3 (message):                                                                                                                                                                                                                                 
  Not building memcpy-bench due to PIC enabled or not Linux or not x86                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                
-- Configuring done (157.7s)                                                                                                                                                                                                                                                                    
-- Generating done (3.1s)                                                                                                                                                                                                                                                                       
CMake Warning:                                                                                                                                                                                                                                                                                  
  Manually-specified variables were not used by the project:                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                
    NO_WERROR 

```

than 
cmake --build build 
```


[0/2] Re-checking globbed directories...
[2815/9144] Linking CXX executable contrib/google-protobuf-cmake/protoc
FAILED: contrib/google-protobuf-cmake/protoc 
: && /usr/bin/c++ -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion -fsized-deallocation  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -ffile-prefix-map=/usr/work/ClickHouse=. -falign-functions=32 -mbranches-within-32B-boundaries -ffp-contract=off  -fdiagnostics-absolute-paths -fstrict-vtable-pointers -w -ffunction-sections -fdata-sections -O2 -g -DNDEBUG -O3 -g  contrib/google-protobuf-cmake/CMakeFiles/protoc.dir/__/google-protobuf/src/google/protobuf/compiler/main.cc.o -o contrib/google-protobuf-cmake/protoc  contrib/google-protobuf-cmake/lib_libprotoc.a  contrib/google-protobuf-cmake/lib_libprotobuf.a  -lpthread  contrib/google-protobuf-cmake/libutf8_validity.a  contrib/abseil-cpp-cmake/libabsl_base.a  contrib/abseil-cpp-cmake/libabsl_cord.a  contrib/abseil-cpp-cmake/libabsl_die_if_null.a  contrib/abseil-cpp-cmake/libabsl_flags.a  contrib/abseil-cpp-cmake/libabsl_hash.a  contrib/abseil-cpp-cmake/libabsl_log_initialize.a  contrib/abseil-cpp-cmake/libabsl_log_severity.a  contrib/abseil-cpp-cmake/libabsl_status.a  contrib/abseil-cpp-cmake/libabsl_statusor.a  contrib/abseil-cpp-cmake/libabsl_strings.a  contrib/abseil-cpp-cmake/libabsl_synchronization.a  contrib/abseil-cpp-cmake/libabsl_time.a  contrib/zlib-ng-cmake/lib_zlib.a  contrib/abseil-cpp-cmake/libabsl_log_internal_check_op.a  contrib/abseil-cpp-cmake/libabsl_leak_check.a  contrib/abseil-cpp-cmake/libabsl_log_internal_conditions.a  contrib/abseil-cpp-cmake/libabsl_log_internal_message.a  contrib/abseil-cpp-cmake/libabsl_log_internal_nullguard.a  contrib/abseil-cpp-cmake/libabsl_examine_stack.a  contrib/abseil-cpp-cmake/libabsl_log_internal_format.a  contrib/abseil-cpp-cmake/libabsl_log_internal_proto.a  contrib/abseil-cpp-cmake/libabsl_log_internal_log_sink_set.a  contrib/abseil-cpp-cmake/libabsl_log_sink.a  contrib/abseil-cpp-cmake/libabsl_log_entry.a  contrib/abseil-cpp-cmake/libabsl_flags_internal.a  contrib/abseil-cpp-cmake/libabsl_flags_marshalling.a  contrib/abseil-cpp-cmake/libabsl_flags_reflection.a  contrib/abseil-cpp-cmake/libabsl_flags_config.a  contrib/abseil-cpp-cmake/libabsl_flags_program_name.a  contrib/abseil-cpp-cmake/libabsl_flags_private_handle_accessor.a  contrib/abseil-cpp-cmake/libabsl_flags_commandlineflag.a  contrib/abseil-cpp-cmake/libabsl_flags_commandlineflag_internal.a  contrib/abseil-cpp-cmake/libabsl_log_globals.a  contrib/abseil-cpp-cmake/libabsl_log_internal_globals.a  contrib/abseil-cpp-cmake/libabsl_raw_hash_set.a  contrib/abseil-cpp-cmake/libabsl_hash.a  contrib/abseil-cpp-cmake/libabsl_city.a  contrib/abseil-cpp-cmake/libabsl_low_level_hash.a  contrib/abseil-cpp-cmake/libabsl_hashtablez_sampler.a  contrib/abseil-cpp-cmake/libabsl_status.a  contrib/abseil-cpp-cmake/libabsl_cord.a  contrib/abseil-cpp-cmake/libabsl_cordz_info.a  contrib/abseil-cpp-cmake/libabsl_cord_internal.a  contrib/abseil-cpp-cmake/libabsl_cordz_functions.a  contrib/abseil-cpp-cmake/libabsl_exponential_biased.a  contrib/abseil-cpp-cmake/libabsl_cordz_handle.a  contrib/abseil-cpp-cmake/libabsl_crc_cord_state.a  contrib/abseil-cpp-cmake/libabsl_crc32c.a  contrib/abseil-cpp-cmake/libabsl_crc_internal.a  contrib/abseil-cpp-cmake/libabsl_crc_cpu_detect.a  contrib/abseil-cpp-cmake/libabsl_bad_optional_access.a  contrib/abseil-cpp-cmake/libabsl_strerror.a  contrib/abseil-cpp-cmake/libabsl_str_format_internal.a  contrib/abseil-cpp-cmake/libabsl_synchronization.a  contrib/abseil-cpp-cmake/libabsl_stacktrace.a  contrib/abseil-cpp-cmake/libabsl_symbolize.a  contrib/abseil-cpp-cmake/libabsl_debugging_internal.a  contrib/abseil-cpp-cmake/libabsl_demangle_internal.a  contrib/abseil-cpp-cmake/libabsl_graphcycles_internal.a  contrib/abseil-cpp-cmake/libabsl_kernel_timeout_internal.a  contrib/abseil-cpp-cmake/libabsl_malloc_internal.a  contrib/abseil-cpp-cmake/libabsl_time.a  contrib/abseil-cpp-cmake/libabsl_strings.a  contrib/abseil-cpp-cmake/libabsl_strings_internal.a  contrib/abseil-cpp-cmake/libabsl_string_view.a  contrib/abseil-cpp-cmake/libabsl_base.a  contrib/abseil-cpp-cmake/libabsl_spinlock_wait.a  contrib/abseil-cpp-cmake/libabsl_throw_delegate.a  contrib/abseil-cpp-cmake/libabsl_int128.a  contrib/abseil-cpp-cmake/libabsl_civil_time.a  contrib/abseil-cpp-cmake/libabsl_time_zone.a  -pthread  contrib/abseil-cpp-cmake/libabsl_bad_variant_access.a  contrib/abseil-cpp-cmake/libabsl_raw_logging_internal.a  contrib/abseil-cpp-cmake/libabsl_log_severity.a  contrib/libcxx-cmake/libcxx.a  contrib/libcxxabi-cmake/libcxxabi.a  contrib/libunwind-cmake/libunwind.a  -nodefaultlibs libclang_rt.builtins-x86_64.a  -lc -lm -lrt -lpthread && :
c++: error: no such file or directory: 'libclang_rt.builtins-x86_64.a'
[2872/9144] Building CXX object contrib/grpc-cmake/CMakeFiles/grpc_plugin_support.dir/__/grpc/src/compiler/python_generator.cc.o
ninja: build stopped: subcommand failed.

```

file libclang_rt.builtins-x86_64.a exists in system 

```
find /usr -name libclang_rt.builtins-x86_64.a
/usr/local/lib/libclang_rt.builtins-x86_64.a
/usr/local/lib/freebsd/libclang_rt.builtins-x86_64.a
/usr/local/llvm15/lib/clang/15.0.7/lib/freebsd/libclang_rt.builtins-x86_64.a

```

if I copy one of them to build directory in Clickhouse 



I recieve another error
```
cp /usr/local/lib/freebsd/libclang_rt.builtins-x86_64.a build/
root@clickhouse://ClickHouse # cmake --build build                                           
[0/2] Re-checking globbed directories...
[4633/9935] Building CXX object contrib/llvm-project/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir[4634/9935] Building CXX object contrib/llvm-project/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir[4634/9935] Building CXX object contrib/llvm-project/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir[4635/9935] Building CXX object contrib/llvm-project/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir[4635/9935] Building CXX object contrib/llvm-project/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir[4636/9935] Building CXX object contrib/llvm-project/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir[4636/9935] Building CXX object contrib/llvm-project/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir[5922/9935] Building C object contrib/libbcryp...CMakeFiles/_bcrypt.dir/__/libbcrypt/bcrypt.c.o 
FAILED: contrib/libbcrypt-cmake/CMakeFiles/_bcrypt.dir/__/libbcrypt/bcrypt.c.o 
/usr/local/bin/ccache /usr/bin/cc -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -isystem /usr/work/ClickHouse/contrib/libbcrypt -isystem /usr/work/ClickHouse/contrib/llvm-project/libcxxabi/include -isystem /usr/work/ClickHouse/contrib/libunwind/include -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -ffile-prefix-map=/usr/work/ClickHouse=. -falign-functions=32 -mbranches-within-32B-boundaries -ffp-contract=off  -fdiagnostics-absolute-paths -w -ffunction-sections -fdata-sections -O2 -g -DNDEBUG -O3 -g  -std=gnu11   -D OS_FREEBSD -Werror -Wno-deprecated-declarations -Wno-poison-system-directories -MD -MT contrib/libbcrypt-cmake/CMakeFiles/_bcrypt.dir/__/libbcrypt/bcrypt.c.o -MF contrib/libbcrypt-cmake/CMakeFiles/_bcrypt.dir/__/libbcrypt/bcrypt.c.o.d -o contrib/libbcrypt-cmake/CMakeFiles/_bcrypt.dir/__/libbcrypt/bcrypt.c.o -c /usr/work/ClickHouse/contrib/libbcrypt/bcrypt.c
In file included from /usr/work/ClickHouse/contrib/libbcrypt/bcrypt.c:22:
/usr/work/ClickHouse/contrib/libbcrypt/crypt_blowfish/ow-crypt.h:27:14: error: conflicting types for 'crypt_r'
   27 | extern char *crypt_r(__const char *key, __const char *setting, void *data);
      |              ^
/usr/include/unistd.h:500:7: note: previous declaration is here
  500 | char    *crypt_r(const char *, const char *, struct crypt_data *);
      |          ^
1 error generated.
[5979/9935] Building CXX object contrib/llvm-p...Files/LLVMX86CodeGen.dir/X86ISelLowering.cpp.o
ninja: build stopped: subcommand failed.

```
I don't know how to compile Clickhouse on Freebsd and can't found any solution in Google :(

I installed llvm15 and llvm17 

```
uname -a 
FreeBSD clickhouse2 13.3-RELEASE FreeBSD 13.3-RELEASE releng/13.3-n257428-80d2b634ddf0 GENERIC amd64

# pkg info |grep llvm
llvm-15_1,1                    Meta-port for the default version of the LLVM Toolchain
llvm15-15.0.7_10               LLVM and Clang
llvm17-17.0.6_5                LLVM and Clang
```

```
 pkg info -l llvm17-17.0.6_5 | grep libclang_rt | grep 'builtins'
        /usr/local/llvm17/lib/clang/17/lib/i386-portbld-freebsd13.2/libclang_rt.builtins.a
        /usr/local/llvm17/lib/clang/17/lib/x86_64-portbld-freebsd13.2/libclang_rt.builtins.a

```
```
 # cmake -S . -B build -DPARALLEL_COMPILE_JOBS=50 -DCMAKE_TOOLCHAIN_FILE=cmake/freebsd/toolchain-x86_64.cmake
-- Cross-compiling for target: 
-- Using compiler:
FreeBSD clang version 17.0.6 (https://github.com/llvm/llvm-project.git llvmorg-17.0.6-0-g6009708b4367)
Target: x86_64-unknown-freebsd13.3
Thread model: posix
InstalledDir: /usr/bin
-- Using linker: <default>
-- Using archiver: /usr/bin/llvm-ar
-- Using ranlib: /usr/bin/llvm-ranlib
-- Using install-name-tool: /usr/local/bin/llvm-install-name-tool
-- Using objcopy: /usr/bin/llvm-objcopy
-- Using strip: /usr/bin/llvm-strip
-- Using ccache: /usr/local/bin/ccache (version 3.7.12)
-- Git HEAD commit hash: 7f79d4f6699b081b43b15930c39bad9eed4ba9b5
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   cmake/freebsd/default_libs.cmake
        modified:   contrib/libunwind (new commits)
        modified:   contrib/mariadb-connector-c (modified content)

no changes added to commit (use "git add" and/or "git commit -a")
-- CMAKE_BUILD_TYPE is not set, set to default = RelWithDebInfo
-- CMAKE_BUILD_TYPE: RelWithDebInfo
-- No official build: A checksum hash will not be added to the clickhouse executable
-- -Bprefix: /usr/local/llvm17/lib/clang/17/lib/x86_64-portbld-freebsd13.2
CMake Error at cmake/freebsd/default_libs.cmake:18 (message):
  libclang_rt.builtins had not been found
Call Stack (most recent call first):
  CMakeLists.txt:389 (include)


-- Configuring incomplete, errors occurred!
```

it's confusing...
in LLVM 17 filename without $ARCH (x86_64)
portbld-freebsd13.2

in LLVM 15 filename + $ARCH
libclang_rt.builtins-x86_64.a



```

ls -la /usr/local/llvm17/lib/clang/17/lib/x86_64-portbld-freebsd13.2/libclang_rt.builtins*
-rw-r--r--  1 root  wheel  245958 Apr  9 03:35 /usr/local/llvm17/lib/clang/17/lib/x86_64-portbld-freebsd13.2/libclang_rt.builtins.a

# ls -la /usr/local/llvm15/lib/clang/15.0.7/lib/freebsd/libclang_rt.builtins-x86_64.a
-rw-r--r--  1 root  wheel  244534 Apr  9 03:11 /usr/local/llvm15/lib/clang/15.0.7/lib/freebsd/libclang_rt.builtins-x86_64.a

--- a/cmake/freebsd/default_libs.cmake
+++ b/cmake/freebsd/default_libs.cmake
@@ -1,10 +1,22 @@
 set (DEFAULT_LIBS "-nodefaultlibs")
 
 if (${CMAKE_SYSTEM_PROCESSOR} STREQUAL "amd64")
-    execute_process (COMMAND ${CMAKE_CXX_COMPILER} --print-file-name=libclang_rt.builtins-x86_64.a OUTPUT_VARIABLE BUILTINS_LIBRARY OUTPUT_STRIP_TRAILING_WHITESPACE)
+    set(system_processor "x86_64")
 else ()
-    execute_process (COMMAND ${CMAKE_CXX_COMPILER} --print-file-name=libclang_rt.builtins-${CMAKE_SYSTEM_PROCESSOR}.a OUTPUT_VARIABLE BUILTINS_LIBRARY OUTPUT_STRIP_TRAILING_WHITESPACE)
+    set(system_processor "${CMAKE_SYSTEM_PROCESSOR}")
 endif ()
+
+file(GLOB bprefix "/usr/local/llvm${COMPILER_VERSION_MAJOR}/lib/clang/${COMPILER_VERSION_MAJOR}/lib/${system_processor}-portbld-freebsd*/")
+message(STATUS "-Bprefix: ${bprefix}")
+
+execute_process(COMMAND ${CMAKE_CXX_COMPILER} -Bprefix=${bprefix} --print-file-name=libclang_rt.builtins-${system_processor}.a OUTPUT_VARIABLE BUILTINS_LIBRARY OUTPUT_STRIP_TRAILING_WHITESPACE)
+# --print-file-name simply prints what you passed in case of nothing was resolved, so let's try one other possible option
+if (BUILTINS_LIBRARY STREQUAL "libclang_rt.builtins-${system_processor}.a")
+    execute_process(COMMAND ${CMAKE_CXX_COMPILER} -Bprefix=${bprefix} --print-file-name=libclang_rt.builtins.a OUTPUT_VARIABLE BUILTINS_LIBRARY OUTPUT_STRIP_TRAILING_WHITESPACE)
+endif()
+if (BUILTINS_LIBRARY STREQUAL "libclang_rt.builtins.a")
+    message(FATAL_ERROR "libclang_rt.builtins had not been found")
+endif()
 
 set (DEFAULT_LIBS "${DEFAULT_LIBS} ${BUILTINS_LIBRARY} ${COVERAGE_OPTION} -lc -lm -lrt -lpthread")
 

